contracts/governance/Vesting/fouryear/FourYearVestingLogic.sol

```diff
@@ -0,0 +1,305 @@
+		uint256 restartDate;
+		uint256 relativeAmount;
+		uint256 periods;
+		if (startDate == 0 && _restartStakeSchedule == 0) {
+			startDate = staking.timestampToLockDate(block.timestamp);
+			durationLeft = duration;
+			cliffAdded = cliff;
+		}
+		if (_restartStakeSchedule > 0 && _restartStakeSchedule == lastStakingSchedule && _amount == remainingStakeAmount) {
+			restartDate = _restartStakeSchedule;
+		} else {
+			restartDate = startDate;
+		}
+		if (endDate == 0) {
+			endDate = staking.timestampToLockDate(block.timestamp.add(duration));
+		}
+		uint256 addedMaxDuration = restartDate.add(maxDuration);
+		if (addedMaxDuration < endDate) {
+			// Runs for max duration
+			lastStakingSchedule = addedMaxDuration;
+			periods = (lastStakingSchedule.sub(restartDate)).div(cliff);
+			uint256 actualSchedule = restartDate.add(periods.mul(cliff));
+			lastStakingSchedule = actualSchedule <= lastStakingSchedule ? actualSchedule : lastStakingSchedule;
+			relativeAmount = (_amount.mul(periods).mul(cliff)).div(durationLeft);
+			durationLeft = durationLeft.sub(periods.mul(cliff));
+			remainingStakeAmount = _amount.sub(relativeAmount);
+		} else {
+			// Normal run
+			lastStakingSchedule = endDate;
+			remainingStakeAmount = 0;
+			durationLeft = 0;
+			relativeAmount = _amount;
+		}
+
+		/// @dev Transfer the tokens to this contract.
+		bool success = SOV.transferFrom(_sender, address(this), relativeAmount);
+		require(success, "transfer failed");
+
+		/// @dev Allow the staking contract to access them.
+		SOV.approve(address(staking), relativeAmount);
+
+		staking.stakesBySchedule(relativeAmount, cliffAdded, duration.sub(durationLeft), cliff, address(this), tokenOwner);
+		if (durationLeft == 0) {
+			cliffAdded = 0;
+		} else {
+			cliffAdded = cliffAdded.add(periods.mul(cliff));
+		}
+
+		emit TokensStaked(_sender, relativeAmount);
+		return (lastStakingSchedule, remainingStakeAmount);
+	}
+
+	/**
+	 * @notice Withdraws tokens from the staking contract and forwards them
+	 * to an address specified by the token owner. Low level function.
+	 * @dev Once here the caller permission is taken for granted.
+	 * @param receiver The receiving address.
+	 * @param isGovernance Whether all tokens (true)
+	 * or just unlocked tokens (false).
+	 * */
+	function _withdrawTokens(address receiver, bool isGovernance) internal {
+		require(receiver != address(0), "receiver address invalid");
+
+		uint96 stake;
+
+		/// @dev Usually we just need to iterate over the possible dates until now.
+		uint256 end;
+
+		/// @dev In the unlikely case that all tokens have been unlocked early,
+		///   allow to withdraw all of them.
+		if (staking.allUnlocked() || isGovernance) {
+			end = endDate;
+		} else {
+			end = block.timestamp;
+			// For four year vesting withdrawal of stakes for the first year is not allowed. These
+			// stakes are extended for three years.
+			require(end > startDate.add(52 weeks), "cannot withdraw in the first year");
```

Comment:
this will only do the job if we extended the duration of all stakes with stakedUntil < startDate.add(52 weeks)
If for some reason, our cron job paused and did not do the extension, the token owner could here withdraw these tokens.

Reply:
Removed the check